### PR TITLE
provider/azurerm: network interfaces and public IP support.

### PR DESCRIPTION
**Pull request by aznashwan**
:exclamation: Not to be merged yet! :exclamation: 

Considering the two resource go hand in hand so well together, this PR aims to bring support for both ARM network interfaces and public IPs.

@phinze @jen20: do please glance over the observations I'll leave on the diff; some of them address your proposal from back off load balancer PR (https://github.com/hashicorp/terraform/pull/4343).

PS: for general continuity; this PR was branched from the local network gateways PR (https://github.com/hashicorp/terraform/pull/4342), so please feel free to ignore those files.

### CODE REVIEW

File: builtin/providers/azurerm/resource_arm_network_interface.go [hunk 1]
```diff
@@ -0,0 +1,393 @@
...
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"network_security_group_id": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+
+			"ip_config": &schema.Schema{
```
Comment (aznashwan) [inline builtin/providers/azurerm/resource_arm_network_interface.go]: @phinze: concerning your suggestions to move some of these sub-resources to the top-level: in this particular case, and I'd suspect many more, Creating a network interface without defining at least one such IP configuration leads to a 400.

Also, as I'll point out in the test config; most of these sub-resources have `id` fields as well and can/must be used in conjunction with other resource/sub-resources.
As such; we may be able to get away with separating some resources; but most will likely turn out impractical to separate.

File: builtin/providers/azurerm/resource_arm_network_interface.go [hunk 2]
```diff
@@ -0,0 +1,393 @@
...
+
+// resourceArmNetworkInterfaceCreate goes ahead and creates the specified ARM network interface.
+func resourceArmNetworkInterfaceCreate(d *schema.ResourceData, meta interface{}) error {
+	ifaceClient := meta.(*ArmClient).ifaceClient
+
+	name := d.Get("name").(string)
+	location := d.Get("location").(string)
+	resGrp := d.Get("resource_group_name").(string)
+	vmId := d.Get("vm_id").(string)
+	// TODO: netSecGrp := d.Get("network_security_group_id").(string)
```
Comment (aznashwan) [inline builtin/providers/azurerm/resource_arm_network_interface.go]: NOTE for when we define security groups...

File: builtin/providers/azurerm/resource_arm_network_interface.go [hunk 3]
```diff
@@ -0,0 +1,393 @@
...
+		}
+	}
+
+	resp, err := ifaceClient.CreateOrUpdate(resGrp, name, network.Interface{
+		Name:     &name,
+		Location: &location,
+		Properties: &network.InterfacePropertiesFormat{
+			VirtualMachine:   &network.SubResource{&vmId},
+			IPConfigurations: &ipconfigs,
+			// TODO: NetworkSecurityGroup: &network.SubResource{&netSecGrp},
```
Comment (aznashwan) [inline builtin/providers/azurerm/resource_arm_network_interface.go]: NOTE for when we define security groups...

File: builtin/providers/azurerm/resource_arm_network_interface.go [hunk 4]
```diff
@@ -0,0 +1,393 @@
...
+				Computed: true,
+			},
+
+			"network_security_group_id": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+
+			"ip_config": &schema.Schema{
+				Type:     schema.TypeList,
```
Comment (aznashwan) [inline builtin/providers/azurerm/resource_arm_network_interface.go]: Also, please note that the existence of this `id` field for these mandatorily nested sub-resources makes using a *schema.Set here lacking a lot in usability.
As such; I'd argue that the ordered indexing of *schema.TypeList is needed here...

File: builtin/providers/azurerm/resource_arm_network_interface.go [hunk 5]
```diff
@@ -0,0 +1,393 @@
...
+	id, err := parseAzureResourceID(d.Id())
+	if err != nil {
+		return fmt.Errorf("Error parsing ID of ARM network interface: %s")
+	}
+	resGrp := id.ResourceGroup
+	name := id.Path["networkInterfaces"]
+
+	// then, issue the actual deletion:
+	resp, err := ifaceClient.Delete(resGrp, name)
+	if resp.StatusCode == http.StatusNotFound {
```
Comment (aznashwan) [inline builtin/providers/azurerm/resource_arm_network_interface.go]: @phinze: wanted confirmation that if the instance got deleted already when we issue one if it's an error or a lucky break?

File: builtin/providers/azurerm/resource_arm_network_interface_test.go [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+}
+
+# TODO: resource "azurerm_instance" "test" ...
+
+resource "azurerm_public_ip" "test" {
+	name = "testAccPublicIPAddress1"
+	resource_group_name = "${azurerm_resource_group.test.name}"
+	location = "${azurerm_resource_group.test.name}"
+	dns_name = "testAccDnsName1"
+	ip_config_id = "${azurerm_network_interface.test.ip_config.0.id}"
```
Comment (aznashwan) [inline builtin/providers/azurerm/resource_arm_network_interface_test.go]: @phinze: my point on using lists for easier accessing of sub-fields.

File: builtin/providers/azurerm/resource_arm_network_interface_test.go [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+}
+
+resource "azurerm_network_interface" "test" {
+    resource_group_name = "${azurerm_resource_group.test.name}"
+	name = "acceptanceTestPublicIPAddress1"
+	location = "West US"
+	vm_id = "${azurerm_instance.test.id}"
+	# TODO: network_security_group_id = ...
```
Comment (aznashwan) [inline builtin/providers/azurerm/resource_arm_network_interface_test.go]: NOTE: for when we implement security groups & rules...

File: builtin/providers/azurerm/resource_arm_public_ip.go [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+	id, err := parseAzureResourceID(d.Id())
+	if err != nil {
+		return fmt.Errorf("Error parsing public IP resource ID: %s", err)
+	}
+	name := id.Path["publicIPAddresses"]
+	resGrp := id.ResourceGroup
+
+	// issue the actual deletion:
+	resp, err := publicIPClient.Delete(resGrp, name)
+	if resp.StatusCode == http.StatusNotFound {
```
Comment (aznashwan) [inline builtin/providers/azurerm/resource_arm_public_ip.go]: again, the question of 404 being a satisfactory deletion...

File: builtin/providers/azurerm/resource_arm_virtual_network.go [hunk 9]
```diff
@@ -45,6 +45,10 @@ func resourceArmVirtualNetwork() *schema.Resource {
 				Required: true,
 				Elem: &schema.Resource{
 					Schema: map[string]*schema.Schema{
+						"id": &schema.Schema{
```
Comment (aznashwan) [inline builtin/providers/azurerm/resource_arm_virtual_network.go]: @phinze, @jen20: note that I added tracking for the ID for inner-vn-subnets for referencing as well...

File: website/source/docs/providers/azurerm/r/public_ip.html.markdown [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+    resource_group_name = "${azurerm_resource_group.test.name}"
+    name = "acceptanceTestPublicIPAddress1"
+    location = "West US"
+    vm_id = "${azurerm_instance.test.id}"
+    # TODO: network_security_group_id = ...
+
+    ip_config = {
+        name = "acceptanceTestIpConfiguration1"
+        dynamic_private_ip = true
+        # TODO: subnet_id = "${azurerm_virtual_network.test.subnet.HASH.id}"
```
Comment (aznashwan) [inline website/source/docs/providers/azurerm/r/public_ip.html.markdown]: NOTE for when we add subnets...

File: website/source/docs/providers/azurerm/r/public_ip.html.markdown [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+    dns_name = "testAccDnsName1"
+    ip_config_id = "${azurerm_network_interface.test.ip_config.0.id}"
+}
+
+resource "azurerm_network_interface" "test" {
+    resource_group_name = "${azurerm_resource_group.test.name}"
+    name = "acceptanceTestPublicIPAddress1"
+    location = "West US"
+    vm_id = "${azurerm_instance.test.id}"
+    # TODO: network_security_group_id = ...
```
Comment (aznashwan) [inline website/source/docs/providers/azurerm/r/public_ip.html.markdown]: NOTE for when we add secgroups...

File: website/source/docs/providers/azurerm/r/public_ip.html.markdown [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+    address_space = ["10.0.0.0/16"]
+    location = "West US"
+
+    subnet {
+        name = "subnet1"
+        address_prefix = "10.0.1.0/24"
+    }
+}
+
+# TODO: resource "azurerm_instance" "test" ...
```
Comment (aznashwan) [inline website/source/docs/providers/azurerm/r/public_ip.html.markdown]: TODO instance declaration for the network interface and public IP.

File: website/source/docs/providers/azurerm/r/network_interface.html.markdown [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+    resource_group_name = "${azurerm_resource_group.test.name}"
+    name = "acceptanceTestPublicIPAddress1"
+    location = "West US"
+    vm_id = "${azurerm_instance.test.id}"
+    # TODO: network_security_group_id = ...
+
+    ip_config = {
+        name = "acceptanceTestIpConfiguration1"
+        dynamic_private_ip = true
+        # TODO: subnet_id = "${azurerm_virtual_network.test.subnet.HASH.id}"
```
Comment (aznashwan) [inline website/source/docs/providers/azurerm/r/network_interface.html.markdown]: NOTE for when we add subnets...

File: website/source/docs/providers/azurerm/r/network_interface.html.markdown [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+    dns_name = "testAccDnsName1"
+    ip_config_id = "${azurerm_network_interface.test.ip_config.0.id}"
+}
+
+resource "azurerm_network_interface" "test" {
+    resource_group_name = "${azurerm_resource_group.test.name}"
+    name = "acceptanceTestPublicIPAddress1"
+    location = "West US"
+    vm_id = "${azurerm_instance.test.id}"
+    # TODO: network_security_group_id = ...
```
Comment (aznashwan) [inline website/source/docs/providers/azurerm/r/network_interface.html.markdown]: NOTE for when we add secgroups...

File: builtin/providers/azurerm/resource_arm_network_interface_test.go [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+
+	ip_config = {
+		name = "acceptanceTestIpConfiguration1"
+		dynamic_private_ip = true
+		# TODO: subnet_id = "${azurerm_virtual_network.test.subnet.HASH.id}"
```
Comment (aznashwan) [inline builtin/providers/azurerm/resource_arm_network_interface_test.go]: TODO: subnets

File: builtin/providers/azurerm/resource_arm_network_interface_test.go [hunk 16]
```diff
@@ -0,0 +1,138 @@
...
+    address_space = ["10.0.0.0/16"]
+    location = "West US"
+
+    subnet {
+        name = "subnet1"
+        address_prefix = "10.0.1.0/24"
+    }
+}
+
+# TODO: resource "azurerm_instance" "test" ...
```
Comment (aznashwan) [inline builtin/providers/azurerm/resource_arm_network_interface_test.go]: TODO: instance declaration to support deployment.

**Comment by aznashwan** [comment]
See you on the other side boys :smile:

File: builtin/providers/azurerm/resource_arm_network_interface.go (hunk 5 above)
Comment (jen20) [inline builtin/providers/azurerm/resource_arm_network_interface.go]: Correct - if we're aiming to delete and it doesn't exist we consider that "success".

**Comment by jen20** [comment]
Hi all, @stack72 ended up implementing these resources #4565 and #4598, which were subsequently merged, so I'll close this pull request now. Thanks for your work @aznashwan!

**Pull request closed without merging**